velox/connectors/hive/HiveDataSource.cpp

```diff
@@ -579,62 +579,49 @@ void HiveDataSource::addSplit(std::shared_ptr<ConnectorSplit> split) {
   }
 
   auto& fileType = reader_->rowType();
+  std::vector<TypePtr> fileColumnTypes = fileType->children();
 
-  std::vector<TypePtr> columnTypes = fileType->children();
-  for (int i = 0; i < readerOutputType_->size(); i++) {
-    auto fieldName = readerOutputType_->nameOf(i);
-    auto scanChildSpec = scanSpec_->childByName(fieldName);
+  auto childrenSpecs = scanSpec_->children();
+  for (int i = 0; i < childrenSpecs.size(); i++) {
+    auto childSpec = childrenSpecs[i].get();
+    std::string fieldName = childSpec->fieldName();
 
-    auto keyIt = split_->partitionKeys.find(fieldName);
-    if (keyIt != split_->partitionKeys.end()) {
-      setPartitionValue(scanChildSpec, fieldName, keyIt->second);
+    auto iter = split_->partitionKeys.find(fieldName);
+    if (iter != split_->partitionKeys.end()) {
+      setPartitionValue(childSpec, fieldName, iter->second);
     } else if (fieldName == kPath) {
-      setConstantValue(
-          scanChildSpec, VARCHAR(), velox::variant(split_->filePath));
+      setConstantValue(childSpec, VARCHAR(), velox::variant(split_->filePath));
     } else if (fieldName == kBucket) {
       if (split_->tableBucketNumber.has_value()) {
         setConstantValue(
-            scanChildSpec,
+            childSpec,
             INTEGER(),
             velox::variant(split_->tableBucketNumber.value()));
       }
-    } else if (!fileType->containsChild(fieldName)) {
-      // Column is missing. Most likely due to schema evolution.
-      setNullConstantValue(scanChildSpec, readerOutputType_->childAt(i));
     } else {
-      // We know the fieldName exists in the file, make the type at that
-      // position match what we expect in the output.
-      columnTypes[fileType->getChildIdx(fieldName)] =
-          readerOutputType_->childAt(i);
-      scanChildSpec->setConstantValue(nullptr);
-    }
-  }
-
-  // Set constant values for partition keys and $path column. If these are
-  // used in filters only, the loop above will miss them.
-  for (const auto& entry : split_->partitionKeys) {
-    auto childSpec = scanSpec_->childByName(entry.first);
-    if (childSpec) {
-      setPartitionValue(childSpec, entry.first, entry.second);
+      auto outputTypeIdx = readerOutputType_->getChildIdxIfExists(fieldName);
```

Comment:
in which case fieldName doesn't exist in readerOutputType_? Is it a valid case or shall we add a check here? thanks!

Reply:
> in which case fieldName doesn't exist in readerOutputType_? Is it a valid case or shall we add a check here? thanks!

For example, some fields/columns are used in the filter but not projected out. For such cases, these fields/columns need to be read to evaluate filters for not one of the readerOutputType_.